internal/reports/report_v2.go

```diff
@@ -0,0 +1,296 @@
+}
+
+type DataReport struct {
+	Rating    string `json:"rating"`
+	BadRows   int64  `json:"badRows"`
+	TotalRows int64  `json:"totalRows"`
+	DryRun    bool   `json:"dryRun"`
+}
+
+type TableReport struct {
+	SrcTableName  string       `json:"srcTableName"`
+	SpTableName   string       `json:"spTableName"`
+	SchemaReport  SchemaReport `json:"schemaReport"`
+	DataReport    DataReport   `json:"dataReport"`
+	Warnings      []Warnings   `json:"warnings"`
+}
+
+type UnexpectedCondition struct {
+	Count     int64  `json:"count"`
+	Condition string `json:"condition"`
+}
+
+type UnexpectedConditions struct {
+	Reparsed             int64
+	UnexpectedConditions []UnexpectedCondition `json:"unexpectedConditions"`
+}
+
+type StructuredReport struct {
+	Summary              Summary              `json:"summary"`
+	IgnoredStatements    []IgnoredStatement   `json:"ignoredStatements"`
+	ConversionMetadata   []ConversionMetadata `json:"conversionMetadata"`
+	MigrationType        string               `json:"migrationType"`
+	StatementStats       StatementStats       `json:"statementStats"`
+	NameChanges          []NameChange         `json:"nameChanges"`
+	TableReports         []TableReport        `json:"tableReports"`
+	UnexpectedConditions UnexpectedConditions `json:"unexpectedConditions"`
+	SchemaOnly           bool                 `json:"-"`
+}
+
+// A report consists of the following parts:
+// 1. Summary (overall quality of conversion)
+// 2. Ignored statements
+// 3. Conversion duration
+// 4. Migration Type
+// 5. Statement stats (in case of dumps)
+// 6. Name changes
+// 7. Individual table reports (Detailed + Quality of conversion for each)
+// 8. Unexpected conditions
+//
+// This method the RAW structured report in JSON format. Several utilities can be built on top of
+// this raw, nested JSON data to output the reports in different user and machine friendly formats
+// such as CSV, TXT etc.
+func GenerateStructuredReport(driverName string, conv *internal.Conv, badWrites map[string]int64, printTableReports bool, printUnexpecteds bool) StructuredReport {
+	//Create report object
+	var hbReport = StructuredReport{}
+	tableReports := AnalyzeTables(conv, badWrites)
+	//1. Generate summary
+	rating, summary := GenerateSummary(conv, tableReports, badWrites)
+	hbReport.Summary = Summary{Text: summary, Rating: rating}
+
+	//2. Ignored Statements
+	hbReport.IgnoredStatements = fetchIgnoredStatements(conv)
+
+	//3. Conversion Metadata
+	hbReport.ConversionMetadata = append(hbReport.ConversionMetadata, ConversionMetadata{ConversionType: "Schema", Duration: conv.Audit.SchemaConversionDuration})
+	hbReport.ConversionMetadata = append(hbReport.ConversionMetadata, ConversionMetadata{ConversionType: "Data", Duration: conv.Audit.DataConversionDuration})
+
+	//4. Migration Type
+	hbReport.MigrationType = mapMigrationType(*conv.Audit.MigrationType)
+
+	//5. Statement statistics
+	var isDump bool
+	if strings.Contains(driverName, "dump") {
```

Comment:
This parsing should ideally be done in the CMD and passed as a flag. Is this peppered through the code ?

Reply:
Yes, it currently is peppered in the code. The `driverName` is used as a proxy to derive if a dump was used or not, hence only `driverName` is passed to this layer.